src/main/java/io/cryostat/net/web/http/api/beta/graph/EnvironmentNodesFetcher.java

```diff
@@ -68,56 +66,24 @@
 
     @Override
     public List<EnvironmentNode> get(DataFetchingEnvironment environment) throws Exception {
-        Map<String, String> filter = environment.getArgument("filter");
+        FilterInput filter = FilterInput.from(environment);
         EnvironmentNode root = rootNodeFetcher.get(environment);
         List<EnvironmentNode> nodes =
                 recurseFetcher.get(
                         DataFetchingEnvironmentImpl.newDataFetchingEnvironment(environment)
                                 .source(root)
                                 .build());
 
-        if (filter == null) {
-            return nodes;
+        if (filter.contains(FilterInput.Key.NAME)) {
+            String nodeName = filter.get(FilterInput.Key.NAME);
+            nodes = recurse(root, n -> Objects.equals(n.getName(), nodeName));
         }
 
-        if (filter.containsKey("nodeName")) {
-            String nodeName = filter.get("nodeName");
-            List<EnvironmentNode> namedNodes =
-                    recurse(root, n -> Objects.equals(n.getName(), nodeName));
-            if (namedNodes.isEmpty()) {
-                throw new NoSuchElementException(String.format("Node named %s", nodeName));
-            }
-            return namedNodes;
-        }
-
-        if (filter.containsKey("nodeType")) {
-            String nodeType = filter.get("nodeType");
-            List<EnvironmentNode> typedNodes =
-                    recurse(root, n -> n.getNodeType().getKind().equalsIgnoreCase(nodeType));
-            if (typedNodes.isEmpty()) {
-                throw new NoSuchElementException(String.format("Node of type %s", nodeType));
-            }
-            return typedNodes;
+        if (filter.contains(FilterInput.Key.NODE_TYPE)) {
+            String nodeType = filter.get(FilterInput.Key.NODE_TYPE);
+            nodes = recurse(root, n -> n.getNodeType().getKind().equalsIgnoreCase(nodeType));
```

Comment:
> The initial implementations of input objects will only support simple string-equality matches, so ex. { name: "myNode", kind: "JVM" } is an input object that logically means something like node.name.equals("myNode") && node.kind.equals("JVM")

Could you explain how this filters nodes with both `nodeName` `AND` `nodeType`? Wouldn't the filtered `nodes` from line 79 get overwritten here?

Reply:
Ah you're correct, good catch. The idea was to perform the query for all nodes first and then sequentially apply each filter, which has the same result as logically ANDing the filter criteria and applying that, but that's not what is actually implemented here. I'll fix it.